site/app/controllers/grading/ElectronicGraderController.php

```diff
@@ -48,6 +52,104 @@ private function validateDiffViewerType(string $type) {
+        */
+        $number_to_grade = $_POST['number_to_grade'];
+        $gradeable = $this->tryGetGradeable($gradeable_id);
+        if ($gradeable === false) {
+            $this->core->addErrorMessage('Invalid Gradeable!');
+            $this->core->redirect($this->core->buildCourseUrl());
+        }
+        $all_grade_all = false;
+        $order = new GradingOrder($this->core, $gradeable, $this->core->getUser(), true);
+        $student_array = [];
+        $student_list = [];
+        $students = $this->core->getQueries()->getUsersByRegistrationSections($order->getSectionNames());
+        foreach ($students as $student) {
+             $reg_sec = ($student->getRegistrationSection() === null) ? 'NULL' : $student->getRegistrationSection();
+             $sorted_students[$reg_sec][] = $student;
+             array_push($student_list, ['user_id' => $student->getId()]);
+             array_push($student_array, $student->getId());
+        }
+        $number_of_students = count($student_list);
+        if ($number_to_grade > $number_of_students) {
+            $all_grade_all = true;
+        }
+        if ($all_grade_all) {
+            $this->core->getOutput()->renderJSONSuccess("Invalid Number of Students Entered");
+            $final_grading_info = [];
+            for ($grader = 0; $grader < count($student_array); ++$grader) {
+                $peer_array = [];
+                for ($peer = 0; $peer < count($student_array); ++$peer) {
+                    if ($grader === $peer) {
+                        continue;
+                    }
+                    else {
+                        array_push($peer_array, $student_array[$peer]);
+                    }
+                }
+                array_push($final_grading_info, [$student_array[$grader],$peer_array]);
+            }
+            $gradeable->setRandomPeerGradersList($final_grading_info);
+            return JsonResponse::getSuccessResponse($final_grading_info);
+        }
+        $graded_array = $student_array;
+          /*n_array_peers : An Array of arrays that holds information on to be graded peers
+          [ [A,B,C,D,E,F], [E,F,A,B,C,D], [C,D,E,F,A,B] ]
+          A grades C and E and is graded by C and E.
+          */
+        $n_array_peers = [];
+        shuffle($student_array);
+        array_push($n_array_peers, $student_array);
+          /*final_grading_info : An Array with clear structure of grading rules for peer grading
+          [ [A,[C,E]],[B,[F,D]], ...]
+          A grades C and E, B grades F and D ..and so on!
+          */
+        $final_grading_info = [];
+        $max_offset = count($student_array);
+        $offset_array = [];
+        for ($i = 0; $i < $number_to_grade; ++$i) {
+            $random_offset = rand(1, $max_offset);
+            if (in_array($random_offset, $offset_array)) {
+                $random_offset = rand($random_offset + 1, $max_offset);
+            }
+            array_push($offset_array, $random_offset);
+        }
```

Comment:
There are still some ways this could break. For example, if another number that has already been picked is picked again.

To fix, make a `Set` containing the numbers `0` through `max_offset`. Then, for `i` to `number_to_grade`, take a random element out of the set (thereby making it impossible to select again) and push that number into the `offset_array`.

Reply:
Please correct me if I am wrong, it looks to me like:
Initially,
$offset_array = [ ] , Suppose $max_offset is 60
Case of Same Random Numbers:
$random_offset= 5 ->Check if $random_offset is in $offset_array -> It's not there -> Push is to $offset_array.
Next Iteration:
$random_offset= 5 ->Check if $random_offset is in $offset_array -> It's there -> Choose another $random_offset from current $random_offset (5+1 = 6) to $max_offset ->push it to array. 

Every time same random offset is chosen by rand(), it will update random offset to a random value from previous random offset +1 to max_offset.